feat #26: 게시물 리스트 조회 API 구현

SteadyFlow736 opened the pull request:
## 📃 설명

- resolves #26 

## 🔨 작업 내용

1.  게시물 리스트 조회 API 구현하였습니다.
2. Querydsl을 이용해 동적 쿼리, 정렬, 페이징 처리하였습니다.

## 💬 기타 사항

-

SteadyFlow736 committed:
test #26: errorCode -> errorCodeType 변경

- CustomException의 errorCode가 errorCodeType 으로 변경됨에 따라 테스트 일부 코드도 변경

SteadyFlow736 committed:
feature #26: 게시물 조회를 위한 커스텀 리포지토리 구현

- 검색 조건을 받고 해당하는 동적 쿼리를 생성할 수 있도록 Querydsl을 이용
- Querydsl을 사용할 수 있도록 커스텀 리포지토리 구현
- 리포지토리 유닛 테스트 추가

SteadyFlow736 committed:
feature #26: 게시물 조회를 위한 서비스 구현

- PostSearchRequest -> PostSearchCondition 변경 기능 구현

SteadyFlow736 committed:
feature #26: 게시물 조회를 위한 컨트롤러 구현

- 통합테스트 추가
- query parameter config 파일 추가

SteadyFlow736 committed:
feature #26: merge

- merge 이후로 PostControllerTest 테스트가 깨지고 있습니다.

SteadyFlow736 committed:
fix #26: 테스트 깨짐 수정

- https://stackoverflow.com/questions/75310776/missing-h2consoleproperties-bean-in-spring-boot-mockmvc-test

JoonheeJeong commented on src/main/java/com/wanted/teamr/snsfeedintegration/repository/PostRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+        return type == null ? null : post.type.eq(type);
+    }
+
+    private BooleanExpression searchContain(SearchByType searchBy, String search) {
+        if (search == null) {
+            return null;
+        }
+
+        if (searchBy == SearchByType.TITLE) {
+            return post.title.contains(search);
+        }
+
+        if (searchBy == SearchByType.CONTENT) {
+            return post.content.contains(search);
+        }
+
+        return post.title.contains(search).or(post.content.contains(search));
```
queryDSL로 조건 깔끔하게 짜셨네요..!

그런데 궁금한 게 null로 리턴되도 괜찮나요?

SteadyFlow736 replied:
감사합니다 😊
searchContain의 리턴값인 BooleanExpression이 null이면 where 절에서는 해당 조건을 사용하지 않게 됩니다.
저희가 원하는 건 검색어(search 파라미터의 값)가 없다면 검색어로 검색하지 않는 것이니 이 때 null을 리턴하게 하여
where search = ... 부분을 삭제하게 하였습니다.

JoonheeJeong commented on src/main/java/com/wanted/teamr/snsfeedintegration/controller/PostController.java:
```diff
@@ -20,6 +24,11 @@ public PostGetResponse getPost(@PathVariable("postId") Long postId) {
         return postService.getPost(postId);
     }
 
+    @GetMapping("/api/posts")
+    public List<PostGetResponse> getPostList(PostSearchRequest request, Pageable pageable) {
+        return postService.getPostList(request, pageable);
+    }
+
```
이후에 추가적으로 인증된 유저에 관한 코드가 들어가서 컨버터에서 resolveHashtag를 수정하면 되겠군요..!

SteadyFlow736 replied:
네 맞습니다 ㅎㅎ

JoonheeJeong commented on src/main/java/com/wanted/teamr/snsfeedintegration/service/PostService.java:
```diff
@@ -2,30 +2,43 @@
     private final SnsService snsService;
+    private final PostSearchConditionConverter converter;
 
     public PostGetResponse getPost(Long postId) {
         Post post = postRepository.findById(postId)
-                                  .orElseThrow(() -> new CustomException(ErrorCode.POST_NOT_FOUND));
+                .orElseThrow(() -> new CustomException(ErrorCode.POST_NOT_FOUND));
         return PostGetResponse.of(post);
     }
 
+    public List<PostGetResponse> getPostList(PostSearchRequest request, Pageable pageable) {
+        PostSearchCondition condition = converter.convert(request);
+        List<Post> posts = postRepository.search(condition, pageable);
```
`@Transactional(readOnly=true)`는 불필요해서 안 쓰신 건가요?

SteadyFlow736 replied:
getPost는 다른 PR(#37 )에서 "조회 수 1 증가" 때문에 Transactional 을 붙였습니다.
getPostList는 일부러 붙이지 않았습니다.
읽기 쿼리가 한 건만 발생하므로 굳이 읽기 트랜잭션을  붙이지 않아도 될 것이라 생각했습니다. 😀

JoonheeJeong reviewed (approved):
QueryDSL과 Pagable로 전반적으로 깔끔하게 구현하셨네요..! 👍🏼

sohyuneeee commented:
조회 조건이 많이 복잡했는데 적절하게 클래스도 분리하고, queryDSL도 사용하셔서 깔끔하게 잘 하신 것 같습니다! 테스트 조건도 어마어마하네요...! 정말 고생 많으셨습니다...!👍🏻

SteadyFlow736 committed:
feat #26: merge develop

- PostRepositoryTest 깨짐 수정
  - PostRepositoryTest 에서 @Import(QuerydslConfig.class) 추가
- PostRepositoryImpl의 JPAQueryFactory 주입 방법 변경
  - PAQueryFactory를 QuerydslConfig에서 등록한 것을 주입받도록 변경

sohyuneeee reviewed (commented):
수고하셨습니다!!! 👍🏻

sohyuneeee commented on src/main/java/com/wanted/teamr/snsfeedintegration/controller/PostController.java:
수고하셨습니다!!! 👍🏻

sohyuneeee reviewed (commented):
수고하셨습니다!!! 👍🏻

sohyuneeee reviewed (commented):
11111
